test/domain/services/notification_service_test.dart

```diff
@@ -0,0 +1,279 @@
+          '{"type":"new_message","groupId":"target-group","messageId":"msg-1"}',
+        );
+
+        when(mockNotification2.id).thenReturn(2);
+        when(mockNotification2.payload).thenReturn(
+          '{"type":"invites_sync","welcomeId":"welcome-1"}', // No groupId (legacy)
+        );
+
+        when(mockNotification3.id).thenReturn(3);
+        when(mockNotification3.payload).thenReturn(null); // Null payload
+
+        when(mockPlugin.getActiveNotifications()).thenAnswer(
+          (_) async => [mockNotification1, mockNotification2, mockNotification3],
+        );
+        when(mockPlugin.cancel(any)).thenAnswer((_) async {});
+
+        // Test the logic
+        final activeNotifications = await mockPlugin.getActiveNotifications();
+        const targetGroupId = 'target-group';
+        final notificationsToCancel = <int>[];
+
+        for (final notification in activeNotifications) {
+          if (notification.payload != null) {
+            try {
+              final payload = Map<String, dynamic>.from(
+                jsonDecode(notification.payload!) as Map,
+              );
+              if (payload['groupId'] == targetGroupId) {
+                notificationsToCancel.add(notification.id!);
+                await mockPlugin.cancel(notification.id!);
+              }
+            } catch (e) {
+              continue;
+            }
+          }
+        }
+
+        // Verify only notification with matching groupId was cancelled
+        expect(notificationsToCancel, equals([1]));
+        verify(mockPlugin.cancel(1)).called(1);
+        verifyNever(mockPlugin.cancel(2)); // No groupId, not cancelled
+        verifyNever(mockPlugin.cancel(3)); // Null payload, not processed
+      });
+
+      test('should handle empty active notifications list', () async {
+        // Arrange: No active notifications
+        when(mockPlugin.getActiveNotifications()).thenAnswer((_) async => []);
+
+        // Test the logic
+        final activeNotifications = await mockPlugin.getActiveNotifications();
+        const targetGroupId = 'target-group';
+        final notificationsToCancel = <int>[];
+
+        for (final notification in activeNotifications) {
+          if (notification.payload != null) {
+            try {
+              final payload = Map<String, dynamic>.from(
+                jsonDecode(notification.payload!) as Map,
+              );
+              if (payload['groupId'] == targetGroupId) {
+                notificationsToCancel.add(notification.id!);
+                await mockPlugin.cancel(notification.id!);
+              }
+            } catch (e) {
+              continue;
+            }
+          }
+        }
+
+        // Verify no notifications were cancelled
+        expect(notificationsToCancel, isEmpty);
+        verify(mockPlugin.getActiveNotifications()).called(1);
+        verifyNever(mockPlugin.cancel(any));
+      });
+    });
+
+    group('actual service calls', () {
```

Comment:
In my opinion, this are the only group tests that should remain, but testing that something just "returns normally" is not a useful test. The tests  in the group above are not calling the method.